server/re/src/mailMS.cpp

```diff
@@ -47,6 +87,16 @@
  **/
 int msiSendMail( msParam_t* xtoAddr, msParam_t* xsubjectLine, msParam_t* xbody, ruleExecInfo_t* ) {
 
+    if (!enable_msiSendMail) {
+        log_msi::error("{}: msiSendMail is disabled. To enable, call msi_enable_msiSendMail() first.", __func__);
+        return SYS_NOT_ALLOWED;
+    }
+
+    // Reset the option to force the admin to consider all locations where
+    // this microservice is used. This guarantees use is always narrowly scoped.
```

Comment:
does this preclude rodsusers from ever calling msiSendMail directly?    i think it does.

i also think i'm okay with that.

Reply:
Indirectly, yes.

Thinking about this more, the rodsadmin requirement means the MSIs can't be in policy triggered by non-admins. As soon as the enable-MSI is hit, an error will occur due to the lack of rodsadmin privileges.

Hmm.

Reply:
correct... by going admin-only ... we've also effectively 'broken' any existing deployments with no workaround available.

does this suggest we need a server-wide variable, that only admins can set (a kind of 'mode')... and it doesn't get 'reset' after each use?

Reply:
That won't work either due to the enable-MSI being callable via clients. Crap.

Reply:
@trel Yes, a server-wide option is the only other (easy) way I can see this working. An environment variable in server_config.json is pretty easy to do without requiring adjustments to JSON schemas.

I'd like to disallow non-admins from being able to call the MSI via `irule` though.

Reply:
i'm suggesting the enable-MSI be admin-only, and msiSendMail not reset the value.

Reply:
or a server_config.json variable, and msiSendMail depend on that.

Reply:
> i'm suggesting the enable-MSI be admin-only, and msiSendMail not reset the value.

I doubt this is doable. How do you envision the admin enabling things?

> or a server_config.json variable, and msiSendMail depend on that.

I'd like to avoid this since that's a change in the schema. And the sendmail MSIs are deprecated.

Reply:
- no enable-MSI
- new advanced_setting for allow_msiSendMail
- server defaults to false
- msiSendMail check this value, exit early if false

Reply:
No change to server_config.json.template since a schema bump isn't going to happen for this.

The server will default to false if the option isn't defined in server_config.json.

Done.